chips/atmega328p-hal/src/wdt.rs

```diff
@@ -0,0 +1,108 @@
+//! Provides a system reset when a counter reaches a given time-out value.
+//!
+//! # Example
+//! ```
+//! let mut watchdog = Wdt::new(&dp.CPU.mcusr, dp.WDT);
+//! watchdog.disable();
+//! watchdog.start(WatchdogTimeOutPeriod::Ms8000);
+//!
+//! loop {
+//!     watchdog.feed();
+//! }
+//! ```
+
+use avr_device::generic::Reg;
+use avr_hal::hal::watchdog::*;
+
+pub enum WatchdogTimeOutPeriod {
+    Ms16,
+    Ms32,
+    Ms64,
+    Ms125,
+    Ms250,
+    Ms500,
+    Ms1000,
+    Ms2000,
+    Ms4000,
+    Ms8000,
+}
+
+pub struct Wdt<'wdt> {
+    mcu_status_register: &'wdt Reg<crate::atmega328p::cpu::mcusr::MCUSR_SPEC>,
+    peripheral: crate::atmega328p::WDT,
+}
+
+impl<'wdt> Wdt<'wdt> {
+    pub fn new(
+        mcu_status_register: &'wdt Reg<crate::atmega328p::cpu::mcusr::MCUSR_SPEC>,
+        peripheral: crate::atmega328p::WDT,
+    ) -> Self {
+        Wdt {
+            mcu_status_register,
+            peripheral,
+        }
+    }
+}
+
+impl<'wdt> WatchdogEnable for Wdt<'wdt> {
+    type Time = WatchdogTimeOutPeriod;
+
+    fn start<T>(&mut self, period: T)
+    where
+        T: Into<Self::Time>,
+    {
+        // Disable interrupts while starting the watchdog timer
+        avr_hal::avr_device::interrupt::free(|_| {
```

Comment:
Can you elaborate why this is necessary?

Reply:
Interrupts were disabled in the code samples in the data sheet. My understanding is that it is because the watchdog configuration must be changed within 4 cycles of enabling changes (writing 1 to WDE and WDCE).

Reply:
Ok, makes sense!  Please document this a bit more in a comment, so any future reader knows immediately what is going on.  Maybe this text from the data-sheet is enough explanation?

> The sequence for clearing WDE and changing time-out configuration is as follows:
> 1. In the same operation, write a logic one to the Watchdog change enable bit (WDCE) and WDE. A logic one must be written to WDE regardless of the previous value of the WDE bit.
> 2. Within the next four clock cycles, write the WDE and Watchdog prescaler bits (WDP) as desired, but with the WDCE bit cleared. This must be done in one operation.